### PR TITLE
Implement parsing time based on pattern for @ApolloJsonValue

### LINKED ISSUE #49: ApolloJsonValue注解增加部分自定义功能
**现状：**
目前ApolloJsonValue仅支持从portal读取指定配置，没有其他功能支持。在遇到下面场景时显得有些无奈：
假如在配置中有一些日期时间相关的配置，如下：

`{
    "beginTime": "2019-08-21 00:00:00",
    "endTime": "2019-09-05 00:00:00",
    "switch": true
}`

在接收对应的时间按配置时只能用String（如果直接用Date接受，会报错：Failed to parse date ["2019-08-21 00:00:00"]: Invalid time zone indicator ' '），然后在代码中进行格式化。
**建议：**
ApolloJsonValue增加一个 dateFormat ，ApolloJsonValueProcessor在解析式可以用指定的格式去解析

### PR DESCRIPTION
**Pull request by huxleyliau**
## What's the purpose of this PR
Implement parsing time based on pattern for @ApolloJsonValue from #49

## Which issue(s) this PR fixes:
Fixes #49 

## Brief changelog

Make use of Gson's current way of turning a string into a date to make @ApolloJsonValue able to handle time strings according to a given pattern.

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Read the [Contributing Guide](https://github.com/apolloconfig/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit tests to verify the code.
- [x] Run `mvn clean test` to make sure this pull request doesn't break anything.
- [x] Update the [`CHANGES` log](https://github.com/apolloconfig/apollo-java/blob/master/CHANGES.md).

### CODE REVIEW

File: apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloAnnotationProcessor.java [hunk 1]
```diff
@@ -244,15 +251,22 @@ private Object resolvePropertyValue(String beanName, String placeHolder) {
     return propertyValue;
   }
 
-  private Object parseJsonValue(String json, Type targetType) {
+  private Object parseJsonValue(String json, Type targetType, String datePattern) {
     try {
-      return GSON.fromJson(json, targetType);
+      return DATEPATTERN_GSON_MAP.computeIfAbsent(datePattern, this::buildGson).fromJson(json, targetType);
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloAnnotationProcessor.java:256]: Is the `computeIfAbsent` method of `HashMap` thread safe?

Comment (huxleyliau) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloAnnotationProcessor.java:256]: Regarding the `HashMap`, `computeIfAbsent` is not thread-safe. If needed, we can use a `ConcurrentHashMap`. However, is there intense thread contention in `AutoUpdateConfigChangeListener`? Is it necessary to make any change?

Comment (nobodyiam) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloAnnotationProcessor.java:256]: I think there would be concurrent execution situations.

Comment (huxleyliau) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/spring/annotation/ApolloAnnotationProcessor.java:256]: I've already changed it to a `ConcurrentHashMap`.  :smiley:

**Review by nobodyiam — APPROVED** [review]
LGTM

**Pull request merged**